vllm/model_executor/layers/moe.py

```diff
@@ -0,0 +1,354 @@
+from typing import Tuple
+
+import torch
+
+from torch import nn
+import torch.nn.functional as F
+
+import triton
+import triton.language as tl
+
+from vllm._C import ops
+from vllm.model_executor.layers.linear import ReplicatedLinear
+
+from vllm.model_executor.parallel_utils.parallel_state import (
+    get_tensor_model_parallel_rank)
+from vllm.model_executor.parallel_utils.communication_op import (
+    tensor_model_parallel_all_reduce)
+from vllm.model_executor.utils import set_weight_attrs
+
+
+class MoE(nn.Module):
+    """A expert parallel MOE that shards each expert across all ranks.
+
+    Each expert's weights are sharded across all ranks. The forward pass
+    will first expand and group the hidden states by experts, then compute
+    the per-rank MLP output of each expert using grouped gemm, and finally
+    reduce the output across ranks.
+    """
+
+    def __init__(
+        self,
+        num_experts: int,
+        top_k: int,
+        hidden_size: int,
+        intermediate_size: int,
+        tp_size: int,
+    ):
+        super().__init__()
+        self.num_total_experts = num_experts
+        self.top_k = top_k
+        self.hidden_size = hidden_size
+        self.intermediate_size = intermediate_size // tp_size
+
+        self.gate = ReplicatedLinear(self.hidden_size,
+                                     self.num_total_experts,
+                                     bias=False,
+                                     linear_method=None)
+
+        self.w1s = nn.Parameter(
+            torch.rand(self.num_total_experts, self.hidden_size,
+                       self.intermediate_size))
+        self.w2s = nn.Parameter(
+            torch.rand(self.num_total_experts, self.intermediate_size,
+                       self.hidden_size))
+        self.w3s = nn.Parameter(
+            torch.rand(self.num_total_experts, self.hidden_size,
+                       self.intermediate_size))
```

Comment:
Could you add `device=torch.cuda.current_device()` or `device="cuda"` when initializing these tensors?

Reply:
Also, please use `torch.empty` instead of `torch.randn` for faster initialization.